src/Functions/FunctionsLogical.cpp

```diff
@@ -507,10 +508,109 @@ DataTypePtr FunctionAnyArityLogical<Impl, Name>::getReturnTypeImpl(const DataTyp
             : result_type;
 }
 
+template <bool inverted>
+static void applyTernaryLogicImpl(const IColumn::Filter & mask, IColumn::Filter & null_bytemap)
+{
+    for (size_t i = 0; i != mask.size(); ++i)
+    {
+        UInt8 value = mask[i];
+        if constexpr (inverted)
+            value = !value;
+
+        if (null_bytemap[i] && value)
+            null_bytemap[i] = 0;
+    }
+}
+
+template <typename Name>
+static void applyTernaryLogic(const IColumn::Filter & mask, IColumn::Filter & null_bytemap)
+{
+    if (Name::name == NameAnd::name)
+        applyTernaryLogicImpl<true>(mask, null_bytemap);
+    else if (Name::name == NameOr::name)
+        applyTernaryLogicImpl<false>(mask, null_bytemap);
+}
+
+template <typename Impl, typename Name>
+ColumnPtr FunctionAnyArityLogical<Impl, Name>::executeShortCircuit(ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type) const
+{
+    if (Name::name != NameAnd::name && Name::name != NameOr::name)
+        throw Exception("Function " + getName() + " doesn't support short circuit execution", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+    /// Let's denote x_i' = maskedExecute(x_i, mask).
+    /// 1) AND(x_0, x_1, x_2, ..., x_n)
+    /// We will support mask_i = x_0 & x_1 & ... & x_i.
+    /// Base:
+    /// mask_0 is 1 everywhere, x_0' = x_0.
+    /// Iteration:
+    /// mask_i = extractMask(mask_{i - 1}, x_{i - 1}')
+    /// x_i' = maskedExecute(x_i, mask)
+    /// Also we will treat NULL as 1 if x_i' is Nullable
+    /// to support ternary logic.
+    /// The result is mask_n.
+    ///
+    /// 1) OR(x_0, x_1, x_2, ..., x_n)
+    /// We will support mask_i = !x_0 & !x_1 & ... & !x_i.
+    /// mask_0 is 1 everywhere, x_0' = x_0.
+    /// mask = extractMask(mask, !x_{i - 1}')
+    /// x_i' = maskedExecute(x_i, mask)
+    /// Also we will treat NULL as 0 if x_i' is Nullable
+    /// to support ternary logic.
+    /// The result is !mask_n.
+
+    bool inverted = Name::name != NameAnd::name;
+    UInt8 null_value = UInt8(Name::name == NameAnd::name);
+    IColumn::Filter mask(arguments[0].column->size(), 1);
+
+    /// If result is nullable, we need to create null bytemap of the resulting column.
+    /// We will fill it while extracting mask from arguments.
+    std::unique_ptr<IColumn::Filter> nulls;
+    if (result_type->isNullable())
+        nulls = std::make_unique<IColumn::Filter>(arguments[0].column->size(), 0);
+
+    MaskInfo mask_info;
+    for (size_t i = 1; i <= arguments.size(); ++i)
+    {
+        if (inverted)
+            mask_info = extractMaskInplaceWithNulls<true>(mask, arguments[i - 1].column, nulls.get(), null_value);
```

Comment:
I think we can treat `inverted` column like `0, 1, Null -> 1, 0, Null`.
If so, we can always use `null_value = 0` (and then remove it).

In other words, I want to replace code
https://github.com/ClickHouse/ClickHouse/blob/c93e14d07f50383bc5c88d1ed5acfb0df6029169/src/Columns/MaskOperations.cpp#L101-L111
to
```cpp
            if (null_bytemap && (*null_bytemap)[index])
            {
                value = 0;
                if (nulls)
                    (*nulls)[i] = 1;
            }
            else
            {
                value = !!data[index];

                if constexpr (inverted)
                    value = !value;
            }
```

And then remove `null_value` from args comepeltely

Reply:
> I think we can treat `inverted` column like `0, 1, Null -> 1, 0, Null`

Can you explain what do you mean here?

Reply:
I want to say, that now we need to specify `null_value` for nulls. But it seems to me that we can ignore `inverted` flag for Nulls, and always use 0 for a mask.

Also, we can look at it in a little bit different way. Assume that `extractMaskInplaceWithNulls<true>(mask, column, ...) = extractMaskInplaceWithNulls<false>(mask, not(column), ...)` where `not` is our not function:
```
SELECT
    x,
    NOT (arrayJoin([0, 1, NULL]) AS x) AS not_x

Query id: b946dc06-d970-43af-ada3-496b8d3544a9

┌────x─┬─not_x─┐
│    0 │     1 │
│    1 │     0 │
│ ᴺᵁᴸᴸ │  ᴺᵁᴸᴸ │
└──────┴───────┘
```

Reply:
I think we cannot do this, because in AND we use `inverted=false` and `null_value=1` (when we see NULL in AND we should continue to determine the result is either NULL or 0).